go/login/server/server.go

```diff
@@ -0,0 +1,178 @@
+	return a(user, hostID, hostIDType, action)
+}
+
+// LoginServer is a framework for glome-login servers, Creation of the object is managed via
+// the NewLoginServer method. To create one it is necessary to provide an authorizer. Optionally, a series
+// of options can be added at creation.
+type LoginServer struct {
+	// Keys manages the keys used by the server.
+	Keys *KeyManager
+
+	// Unexported fields.
+	auth        Authorizer
+	authLock    sync.RWMutex
+	loginParser *login.Server
+
+	responseLen uint8
+	userHeader  string
+}
+
+// Authorizer sets Authorizer function for LoginServer in a concurrentsafe way.
+func (s *LoginServer) Authorizer(a Authorizer) {
+	s.authLock.Lock()
+	s.auth = a
+	s.authLock.Unlock()
+}
+
+// NewLoginServer creates a new server with provided Authorizer and, optionally, selected options.
+func NewLoginServer(a Authorizer, options ...func(*LoginServer) error) (*LoginServer, error) {
+	srv := LoginServer{
+		auth:        a,
+		Keys:        NewKeyManager(),
+		responseLen: MaxResponseSize,
+	}
+	srv.loginParser = srv.newLoginParser()
+
+	for _, option := range options {
+		if err := option(&srv); err != nil {
+			return nil, err
+		}
+	}
+
+	return &srv, nil
+}
+
+// ResponseLen is an option to be provided to NewServer on creation. Its sets responseLen to
+// provided length. Raise error if provided length is not in 0-MaxResponseSize
+func ResponseLen(length uint8) func(srv *LoginServer) error {
+	return func(srv *LoginServer) error {
+		if !(0 < length && length <= MaxResponseSize) {
+			return ErrInvalidResponseLen
+		}
+		srv.responseLen = length
+		return nil
+	}
+}
+
+// UserHeader is an option to be provided to NewServer on creation. Its sets UserHeader to
+// provided string.
+func UserHeader(s string) func(srv *LoginServer) error {
+	return func(srv *LoginServer) error {
+		srv.userHeader = s
+		return nil
+	}
+}
+
+func (s *LoginServer) newLoginParser() *login.Server {
+	return &login.Server{KeyFetcher: s.Keys.keyFetcher()}
+}
+
+// ServeHTTP implements http.Handler interface:
+// - On "/": List server service keys.
+// - On a glome login URL: Return an login token or an error message.
+func (s *LoginServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path == "/" {
+		s.printServerKeys(w)
+		return
+	}
+
+	user := r.Header.Get(s.userHeader)
+
+	path := r.URL.RawPath
+	if path == "" {
+		path = r.URL.Path
+	}
+
+	response, err := s.loginParser.ParseURLResponse(path)
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+
+	s.printToken(w, response, user)
+}
+
+// Auxiliary function to print login token on response writer.
+func (s *LoginServer) printToken(w http.ResponseWriter, r *login.URLResponse, user string) {
+	s.authLock.RLock()
+	allowed, err := s.auth.GrantLogin(user, r.Msg.HostID, r.Msg.HostIDType,
+		r.Msg.Action)
+	s.authLock.RUnlock()
+
+	if !allowed {
+		if err != nil {
+			http.Error(w, err.Error(), 403)
+		}
+		http.Error(w, "unauthorized action", 403)
```

Comment:
Maybe in an `else` block, if you want to share the `return`?

Reply:
Added